### PR TITLE
Schedule state and callback at the same time

### PR DESCRIPTION
**Pull request by acdlite**
Fixes an issue (in both Stack and Fiber) where `enqueueSetState` causes a synchronous update that flushes before `enqueueCallback` is ever called. Now `enqueueSetState` et al accept an optional callback so that both are scheduled at the same time.

### CODE REVIEW

File: src/renderers/shared/fiber/ReactFiberUpdateQueue.js [hunk 1]
```diff
@@ -367,14 +377,18 @@ exports.getPendingPriority = getPendingPriority;
...
   fiber : Fiber,
   partialState : PartialState<any, any>,
+  callback : Callback | null,
   priorityLevel : PriorityLevel
 ) : void {
-  const isTopLevelUnmount = partialState === null;
+  const isTopLevelUnmount = Boolean(
+    partialState &&
+    partialState.element === null
+  );
```
Comment (bvaughn) [inline src/renderers/shared/fiber/ReactFiberUpdateQueue.js]: At a glance, this change seems unrelated to the core of this PR, and also like the opposite of what the logic used to be. Was it a bug before? I don't see any other change inside of this PR that looks as though it would have invalidated the previous check, but maybe I'm missing it.

File: src/renderers/shared/stack/reconciler/ReactUpdateQueue.js [hunk 2]
```diff
@@ -244,6 +268,15 @@ var ReactUpdateQueue = {
...
     queue.push(partialState);
 
+    if (callback) {
+      ReactUpdateQueue.validateCallback(callback, callerName);
+      if (internalInstance._pendingCallbacks) {
+        internalInstance._pendingCallbacks.push(callback);
+      } else {
+        internalInstance._pendingCallbacks = [callback];
+      }
+    }
```
Comment (bvaughn) [inline src/renderers/shared/stack/reconciler/ReactUpdateQueue.js]: Do we manually inline these 3 blocks (and the one in `enqueueCallback`) for performance reasons?

File: src/renderers/shared/shared/__tests__/ReactCompositeComponentState-test.js [hunk 3]
```diff
@@ -171,39 +171,20 @@ describe('ReactCompositeComponent-state', () => {
...
       ['componentWillUpdate-currentState', 'orange'],
       ['componentWillUpdate-nextState', 'yellow'],
       ['render', 'yellow'],
       ['componentDidUpdate-currentState', 'yellow'],
       ['componentDidUpdate-prevState', 'orange'],
       ['setState-yellow', 'yellow'],
-    );
-
-    if (ReactDOMFeatureFlags.useFiber) {
-      expected.push(['initial-callback', 'yellow']);
```
Comment (spicyj) [inline src/renderers/shared/shared/__tests__/ReactCompositeComponentState-test.js]: What happened here?

File: src/renderers/shared/shared/__tests__/ReactUpdates-test.js [hunk 4]
```diff
@@ -648,16 +648,18 @@ describe('ReactUpdates', () => {
           'Inner-render-1-0',
           'Inner-didUpdate-1-0',
         'Outer-didUpdate-1',
+           // Happens in a batch, so don't re-render yet
           'Inner-setState-1',
-            'Inner-render-1-1',
```
Comment (spicyj) [inline src/renderers/shared/shared/__tests__/ReactUpdates-test.js]: Why did this rerender immediately before? I thought this change would only move some setState callbacks earlier.

File: src/renderers/shared/shared/__tests__/ReactCompositeComponentState-test.js (hunk 3 above)
Comment (acdlite) [inline src/renderers/shared/shared/__tests__/ReactCompositeComponentState-test.js]: Explained by the now-deleted comment above:

> In Fiber, the initial callback is not enqueued until after any work scheduled by lifecycles has flushed (same semantics as a regular setState callback outside of a batch). In Stack, the initial render is scheduled inside of batchedUpdates, so the callback gets flushed right after componentDidMount.
>
> TODO: We should fix this, in both Stack and Fiber, so that the behavior is consistent regardless of whether you're in a batch.

File: src/renderers/shared/shared/__tests__/ReactUpdates-test.js (hunk 4 above)
Comment (acdlite) [inline src/renderers/shared/shared/__tests__/ReactUpdates-test.js]: I'm actually not sure why it re-rendered immediately before, since `setState` is called inside `componentDidUpdate` which should happen in a batch regardless. Is this too significant a change to make in a non-major release?

File: src/renderers/shared/stack/reconciler/ReactUpdateQueue.js (hunk 2 above)
Comment (acdlite) [inline src/renderers/shared/stack/reconciler/ReactUpdateQueue.js]: Didn't seem worth extracting into a function.

File: src/renderers/shared/fiber/ReactFiberUpdateQueue.js (hunk 1 above)
Comment (acdlite) [inline src/renderers/shared/fiber/ReactFiberUpdateQueue.js]: Your suggested change is wrong because

```js
Object.assign({}, { element: element }, null);
```

merely copies the previous state, whereas an unmount sets `element` to null.

Comment (bvaughn) [inline src/renderers/shared/fiber/ReactFiberUpdateQueue.js]: > The previous test was just wrong.

Gotcha. This is what I thought may be the case but I wasn't sure.

File: src/renderers/shared/shared/__tests__/ReactUpdates-test.js (hunk 4 above)
Comment (spicyj) [inline src/renderers/shared/shared/__tests__/ReactUpdates-test.js]: We should figure out why there was a change before we can make a meaningful decision on that point. We've made other breaking changes for Fiber so we'll need to do a 16 regardless before releasing Fiber (or maybe at the same time).

File: src/renderers/shared/fiber/ReactFiberClassComponent.js [hunk 5]
```diff
@@ -31,30 +35,36 @@ var invariant = require('invariant');
...
 ) {
 
   // Class component state updater
   const updater = {
     isMounted,
-    enqueueSetState(instance, partialState) {
+    enqueueSetState(instance, partialState, callback) {
       const fiber = ReactInstanceMap.get(instance);
-      scheduleSetState(fiber, partialState);
+      const priorityLevel = getPriorityContext();
```
Comment (spicyj) [inline src/renderers/shared/fiber/ReactFiberClassComponent.js]: Why did you choose to split these out rather than adding `callback` as an argument to `scheduleSetState` etc? This feels like it's asking for someone to accidentally introduce the problem this PR is fixing.

Comment (acdlite) [inline src/renderers/shared/fiber/ReactFiberClassComponent.js]: I was going to remove `scheduleSetState` et al from the scheduler anyway because they were just an extra layer of indirection around `addUpdate` et al. See @sebmarkbage's comment here for context: https://github.com/facebook/react/pull/8538#discussion_r92308098

To avoid introducing this problem again we could delete `addCallback` since it's no longer used. Sebastian thought it was better not to break the existing class updater API on the off chance someone was relying on it.

Comment (spicyj) [inline src/renderers/shared/fiber/ReactFiberClassComponent.js]: I'd rather remove them. I don't think we've ever described these as public at all. Google doesn't show anyone talking about them. I'll leave it up to you though.

Comment (acdlite) [inline src/renderers/shared/fiber/ReactFiberClassComponent.js]: Ok then I'll remove
